### ISSUE TITLE
SNI not set in client-hello when "test" is used as hostname

### ISSUE THREAD

**Issue by saran-saran007** [openssl/openssl#17735]
I can see the hostname set when using string other that test* in - SSL_set_tlsext_host_name(ssl, <string variable>). But, when I use SSL_set_tlsext_host_name(ssl, **"test"**) the client hello doesn't contain a SNI extension and also results in a successful handshake. Where as my expectation is that "test" should have been sent in SNI extension and resulting in Handshake failure due to wrong hostname.( am using X509_VERIFY_PARAM_set_hostflags and X509_VERIFY_PARAM_set1_host) to force hostname checking.

code:-
```` C
void enable_hostname_verification(SSL *ssl, const char *domain)
{
        X509_VERIFY_PARAM *param = NULL;
        int ret = 0;

        param = SSL_get0_param(ssl);

        /* Enable automatic hostname checks */
        X509_VERIFY_PARAM_set_hostflags(param, X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS);
        if (!X509_VERIFY_PARAM_set1_host(param, domain, sizeof(domain) - 1)) {
                // handle error
                return;
        }

        ret = SSL_set_tlsext_host_name(ssl, domain);
        if (ret != 1)
        {
            printf("### unable to set hostname to: %s ###", domain);
        } else {
            printf("### set SNI  to: %s ###", domain);
        }
        /* Enable peer verification, (with a non-null callback if desired) */
        SSL_set_verify(ssl, SSL_VERIFY_PEER, NULL);
}
````
Issue occurs when I call the above function as enable_hostname_verification(ss, "test").

**Comment by mattcaswell** [comment]
Hmmm. I don't see this behaviour here. Using OpenSSL 3.0.1 if I use the "-servername" option to s_client then that causes s_client to call `SSL_set_tlsext_host_name` with the supplied parameter:

```` C
        if (servername != NULL && !SSL_set_tlsext_host_name(con, servername)) {
            BIO_printf(bio_err, "Unable to set TLS servername extension.\n");
            ERR_print_errors(bio_err);
            goto end;
        }
````

So:

````
$ openssl s_client -servername test -trace
CONNECTED(00000003)
Sent Record
Header:
  Version = TLS 1.0 (0x301)
  Content Type = Handshake (22)
  Length = 305
    ClientHello, Length=301
...sniip...
      extensions, length = 166
        extension_type=server_name(0), length=9
          0000 - 00 07 00 00 04 74 65 73-74                     .....test
````

**Comment by saran-saran007** [comment]
Sorry, my bad. Am able to see the SNI set properly. However, I see  a successful handshake while sending "test" as a SNI even if the client receives a server-certificate with "comman-name" set to "TLS-Unit-test"

**Comment by mattcaswell** [comment]
> the client receives a server-certificate with "comman-name" set to "TLS-Unit-test"

The setting of SNI does not affect the host that the client checks in the certificate. This is set via `X509_VERIFY_PARAM_set1_host` and must be accompanied by a call to `SSL_set_verify` with `SSL_VERIFY_PEER`. In your sample code you do seem to set that. This line looks wrong:

```` C
         if (!X509_VERIFY_PARAM_set1_host(param, domain, sizeof(domain) - 1)) {
````

`sizeof(domain)` here will be the size of the ptr type, not the length of the string. But this probably doesn't explain the behaviour you are seeing.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of the 3.4 dev cycle, barring further input
